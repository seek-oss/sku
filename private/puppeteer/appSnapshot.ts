import type { ExpectStatic } from 'vitest';
import { TEST_TIMEOUT } from '@sku-private/test-utils/constants';

export const getAppSnapshot = async ({
  url,
  warningFilter = () => true,
  expect,
  waitUntil = 'networkidle0',
  timeout = TEST_TIMEOUT,
}: {
  url: string;
  warningFilter?: (warning: string) => boolean;
  expect: ExpectStatic;
  waitUntil?: 'load' | 'domcontentloaded' | 'networkidle0' | 'networkidle2';
  timeout?: number;
}) => {
  const warnings: string[] = [];
  const errors: string[] = [];

  const appPage = await browser.newPage();
  appPage.setDefaultNavigationTimeout(timeout);

  appPage.on('console', (msg) => {
    if (msg.type() === 'warn') {
      warnings.filter(warningFilter).push(msg.text());
    }

    if (msg.type() === 'error') {
      let isFaviconError = false;
      msg.stackTrace().forEach((frame) => {
        // Ignore 404s for favicons
        if (frame?.url?.endsWith('favicon.ico')) {
          isFaviconError = true;
          return;
        }

        errors.push(`${frame.url}:${frame.lineNumber}:${frame.columnNumber}`);
      });

      if (isFaviconError) {
        return;
      }

      errors.push(msg.text());
    }
  });

  try {
    const response = await appPage.goto(url, {
      waitUntil,
    });
    const sourceHtml = (await response?.text()) || '';
    const clientRenderContent = await appPage.content();

    expect(warnings).toEqual([]);
    expect(errors).toEqual([]);

    return { sourceHtml, clientRenderContent };
  } finally {
    await appPage.close();
  }
};
