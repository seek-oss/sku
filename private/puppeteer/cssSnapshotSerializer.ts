import prettier from '@prettier/sync';
import { parse } from 'css';
import type { SnapshotSerializer } from 'vitest';

/** Formats CSS strings */
export const cssSnapshotSerializer: SnapshotSerializer = {
  print: (value) => {
    if (typeof value !== 'string') {
      throw new Error(
        `cssSnapshotSerializer expected a string, received ${typeof value}`,
      );
    }

    return prettier.format(value, { parser: 'css' });
  },
  test: (value) => {
    try {
      parse(value);
    } catch {
      return false;
    }

    return true;
  },
};
