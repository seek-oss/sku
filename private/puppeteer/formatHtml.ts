import diffableHtml from 'diffable-html';

export const formatHtml = (html: string) => diffableHtml(html).trim();
