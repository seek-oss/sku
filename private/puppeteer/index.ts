export { getAppSnapshot } from './appSnapshot.ts';
export { appSnapshotSerializer } from './appSnapshotSerializer.ts';
export { htmlSnapshotSerializer } from './htmlSnapshotSerializer.ts';
export { cssSnapshotSerializer } from './cssSnapshotSerializer.ts';
