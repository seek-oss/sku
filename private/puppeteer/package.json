{
  "name": "@sku-private/puppeteer",
  "private": true,
  "type": "module",
  "main": "index.ts",
  "devDependencies": {
    "@prettier/sync": "^0.6.1",
    "@sku-private/test-utils": "workspace:*",
    "@sku-private/testing-library": "workspace:*",
    "@types/css": "^0.0.38",
    "@types/diffable-html": "^5.0.2",
    "css": "^3.0.0",
    "diff": "^8.0.2",
    "diffable-html": "^6.0.1"
  }
}
