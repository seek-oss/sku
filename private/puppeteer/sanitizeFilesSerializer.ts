import type { SnapshotSerializer } from 'vitest';
import { sanitizeString } from './sanitizeString.ts';

/** Formats CSS strings */
export const sanitizeFilesSerializer = {
  serialize(val: string, config, indentation, depth, refs, printer) {
    // Runs the default serializer on the sanitized value
    return printer(
      sanitizeString(val),
      { ...config, plugins: [] },
      indentation,
      depth,
      refs,
    );
  },

  test: (value) => value && typeof value === 'string',
} satisfies SnapshotSerializer;
