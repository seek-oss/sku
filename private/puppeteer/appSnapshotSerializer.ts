import type { SnapshotSerializer } from 'vitest';
import { createTwoFilesPatch } from 'diff';
import { formatHtml } from './formatHtml.ts';
import { sanitizeString } from './sanitizeString.ts';

const emptyDiff = `===================================================================
--- sourceHtml
+++ clientHtml`;

/**
 * Serializes the output of `getAppSnapshot`, formatting the HTML and diffing the pre and
 * post-hydration HTML
 */
export const appSnapshotSerializer: SnapshotSerializer = {
  serialize: ({ sourceHtml, clientRenderContent }) => {
    const formattedSourceHtml = formatHtml(sourceHtml);
    const formattedClientHtml = formatHtml(clientRenderContent);

    const htmlDiff = createTwoFilesPatch(
      'sourceHtml',
      'clientHtml',
      formattedSourceHtml,
      formattedClientHtml,
      undefined,
      undefined,
      { ignoreNewlineAtEof: true, context: 3 },
    ).trim();

    const isEmptyDiff = htmlDiff === emptyDiff;

    const snapshotItems = [
      formattedSourceHtml,
      `POST HYDRATE DIFFS: ${isEmptyDiff ? 'NO DIFF' : `\n${htmlDiff}`}`,
    ];

    return sanitizeString(snapshotItems.join('\n'));
  },

  test: (val) =>
    val &&
    val.hasOwnProperty('clientRenderContent') &&
    val.hasOwnProperty('sourceHtml'),
};
