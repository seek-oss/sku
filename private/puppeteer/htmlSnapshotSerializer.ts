import { formatHtml } from './formatHtml.ts';

import type { SnapshotSerializer } from 'vitest';

/**
 * Typically used to serialize the result of `dirContentsToObject`. It groups assets by type, and
 * formats CSS and HTML asset contents.
 */
export const htmlSnapshotSerializer: SnapshotSerializer = {
  print: (html, serializer) => {
    const assets: { SCRIPTS: string[]; CSS: string[] } = {
      SCRIPTS: [],
      CSS: [],
    };

    if (typeof html !== 'string') {
      throw new Error(
        `htmlSnapshotSerializer expected a string, received ${typeof html}`,
      );
    }

    const extractedHtml = formatHtml(html).replace(
      /(href|src)="(.*\.(?:js|css))"/g,
      (_match, key, url) => {
        const assetType = url.endsWith('.js') ? 'SCRIPTS' : 'CSS';

        let assetIndex = assets[assetType].indexOf(url);
        if (assetIndex === -1) {
          assetIndex = assets[assetType].push(url) - 1;
        }

        return `${key}="${assetType}[${assetIndex}]"`;
      },
    );

    return [
      `SCRIPTS: ${serializer(assets.SCRIPTS)}`,
      `CSS: ${serializer(assets.CSS)}`,
      `SOURCE HTML: ${formatHtml(extractedHtml)}`,
    ].join('\n');
  },
  test: (value) =>
    typeof value === 'string' && value.startsWith('<!DOCTYPE html>'),
};
