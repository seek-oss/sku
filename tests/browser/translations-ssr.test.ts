import {
  describe,
  beforeAll,
  afterAll,
  it,
  expect as globalExpect,
} from 'vitest';
import { getAppSnapshot } from '@sku-private/puppeteer';
import {
  cleanup,
  scopeToFixture,
  skipCleanup,
} from '@sku-private/testing-library';

const { render } = scopeToFixture('translations');

const backendUrl = `http://localhost:8314`;

describe('ssr translations', () => {
  beforeAll(async () => {
    const startSsr = await render('start-ssr', ['--config=sku-ssr.config.ts']);
    globalExpect(
      await startSsr.findByText('Server started'),
    ).toBeInTheConsole();
  });

  afterAll(cleanup);

  it('should render en', async ({ expect, task }) => {
    skipCleanup(task.id);
    const app = await getAppSnapshot({ expect, url: `${backendUrl}/en` });
    expect(app).toMatchSnapshot();
  });

  it('should render fr', async ({ expect, task }) => {
    skipCleanup(task.id);
    const app = await getAppSnapshot({ expect, url: `${backendUrl}/fr` });
    expect(app).toMatchSnapshot();
  });

  it('should render en-PSEUDO', async ({ expect, task }) => {
    skipCleanup(task.id);
    const app = await getAppSnapshot({
      expect,
      url: `${backendUrl}/en?pseudo=true`,
    });
    expect(app).toMatchSnapshot();
  });
});
