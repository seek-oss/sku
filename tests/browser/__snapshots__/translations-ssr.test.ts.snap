// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ssr translations > should render en 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/runtime.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/5.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/main.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="en-translations"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/en-translations.js"
      crossorigin="anonymous"
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello sku
        Company name
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
      crossorigin="anonymous"
    >
      [1]
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
      crossorigin="anonymous"
    >
      {"namedChunks":["en-translations"]}
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/runtime.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/5.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/main.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="en-translations"
      src="http://127.0.0.1:8313/en-translations.js"
      crossorigin="anonymous"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: NO DIFF
`;

exports[`ssr translations > should render en-PSEUDO 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/runtime.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/5.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/main.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="en-PSEUDO-translations"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/en-PSEUDO-translations.js"
      crossorigin="anonymous"
    >
  </head>
  <body>
    <div id="app">
      <div>
        [Ḩẽẽẽƚƚööö šķǚǚǚ]
        [Çööm̂ƥăăกี้ýý กี้ăăm̂ẽẽ]
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
      crossorigin="anonymous"
    >
      [0]
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
      crossorigin="anonymous"
    >
      {"namedChunks":["en-PSEUDO-translations"]}
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/runtime.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/5.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/main.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="en-PSEUDO-translations"
      src="http://127.0.0.1:8313/en-PSEUDO-translations.js"
      crossorigin="anonymous"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: NO DIFF
`;

exports[`ssr translations > should render fr 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/runtime.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/5.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/main.js"
      crossorigin="anonymous"
    >
    <link
      data-chunk="fr-translations"
      rel="preload"
      as="script"
      href="http://127.0.0.1:8313/fr-translations.js"
      crossorigin="anonymous"
    >
  </head>
  <body>
    <div id="app">
      <div>
        Bonjour sku
        บริษัท
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
      crossorigin="anonymous"
    >
      [2]
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
      crossorigin="anonymous"
    >
      {"namedChunks":["fr-translations"]}
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/runtime.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/5.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="http://127.0.0.1:8313/main.js"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="fr-translations"
      src="http://127.0.0.1:8313/fr-translations.js"
      crossorigin="anonymous"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: NO DIFF
`;
